### PR TITLE
[api] Add API stability annotations for all RLlib symbols and add to LINT

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds programmatic enforcement that all RLlib symbols have API stability annotations. I've based the stability annotations here on inputs from @gjoliver , but they should be audited carefully by @sven1977 and RLlib team since I did them quite quickly.

### CODE REVIEW

File: rllib/utils/framework.py [hunk 1]
```diff
@@ -4,12 +4,14 @@
...
 from typing import Any, Optional
 
+from ray.rllib.utils.annotations import DeveloperAPI
 from ray.rllib.utils.deprecation import Deprecated
 from ray.rllib.utils.typing import TensorShape, TensorType
 
 logger = logging.getLogger(__name__)
 
 
+@DeveloperAPI
```
Comment (sven1977) [inline rllib/utils/framework.py:14]: PublicAPI

File: rllib/utils/framework.py [hunk 2]
```diff
@@ -41,6 +43,7 @@ def try_import_jax(error: bool = False):
     return jax, flax
 
 
+@DeveloperAPI
```
Comment (sven1977) [inline rllib/utils/framework.py:46]: PublicAPI

File: rllib/utils/framework.py [hunk 3]
```diff
@@ -121,6 +125,7 @@ def decorator(func):
     return decorator
 
 
+@DeveloperAPI
```
Comment (sven1977) [inline rllib/utils/framework.py:128]: PublicAPI

File: rllib/utils/exploration/random_encoder.py [hunk 4]
```diff
@@ -16,6 +17,7 @@
 tf1, tf, tfv = try_import_tf()
 
 
+@PublicAPI
```
Comment (sven1977) [inline rllib/utils/exploration/random_encoder.py:20]: _private class

File: rllib/env/wrappers/dm_env_wrapper.py [hunk 5]
```diff
@@ -34,6 +36,7 @@ def _convert_spec_to_space(spec):
     )
 
 
+@PublicAPI
```
Comment (sven1977) [inline rllib/env/wrappers/dm_env_wrapper.py:39]: DeveloperAPI

File: rllib/env/wrappers/dm_control_wrapper.py [hunk 6]
```diff
@@ -71,6 +73,7 @@ def _flatten_obs(obs):
     return np.concatenate(obs_pieces, axis=0)
 
 
+@PublicAPI
```
Comment (sven1977) [inline rllib/env/wrappers/dm_control_wrapper.py:76]: DeveloperAPI

File: rllib/env/wrappers/unity3d_env.py [hunk 7]
```diff
@@ -7,11 +7,13 @@
 
 from ray.rllib.env.multi_agent_env import MultiAgentEnv
 from ray.rllib.policy.policy import PolicySpec
+from ray.rllib.utils.annotations import DeveloperAPI
 from ray.rllib.utils.typing import MultiAgentDict, PolicyID, AgentID
 
 logger = logging.getLogger(__name__)
 
 
+@DeveloperAPI
```
Comment (sven1977) [inline rllib/env/wrappers/unity3d_env.py:16]: PublicAPI

File: rllib/env/wrappers/pettingzoo_env.py [hunk 8]
```diff
@@ -141,6 +143,7 @@ def get_sub_environments(self):
         return self.env.unwrapped
 
 
+@DeveloperAPI
```
Comment (sven1977) [inline rllib/env/wrappers/pettingzoo_env.py:146]: PublicAPI

File: rllib/env/wrappers/pettingzoo_env.py [hunk 9]
```diff
@@ -1,6 +1,8 @@
 from ray.rllib.env.multi_agent_env import MultiAgentEnv
+from ray.rllib.utils.annotations import DeveloperAPI
 
 
+@DeveloperAPI
```
Comment (sven1977) [inline rllib/env/wrappers/pettingzoo_env.py:5]: PublicAPI

**Comment by sven1977** [comment]
Thanks for doing this @ericl ! Looks good, except for a few changes, which I commented on.

**Comment by sven1977** [comment]
Also, we moved to gym==0.23, so there could be some conflicts wrt. the gym rendering wrappers.

File: rllib/env/wrappers/pettingzoo_env.py [hunk 10]
```diff
@@ -1,6 +1,8 @@
 from ray.rllib.env.multi_agent_env import MultiAgentEnv
+from ray.rllib.utils.annotations import DeveloperAPI
```
Comment (sven1977) [inline rllib/env/wrappers/pettingzoo_env.py:2]:
```suggestion
from ray.rllib.utils.annotations import PublicAPI
```

File: rllib/env/wrappers/pettingzoo_env.py (hunk 9 above)
Comment (sven1977) [inline rllib/env/wrappers/pettingzoo_env.py:5]:
```suggestion
@PublicAPI
```

File: rllib/env/wrappers/pettingzoo_env.py (hunk 8 above)
Comment (sven1977) [inline rllib/env/wrappers/pettingzoo_env.py:146]:
```suggestion
@PublicAPI
```

File: rllib/env/wrappers/unity3d_env.py [hunk 11]
```diff
@@ -7,11 +7,13 @@
 
 from ray.rllib.env.multi_agent_env import MultiAgentEnv
 from ray.rllib.policy.policy import PolicySpec
+from ray.rllib.utils.annotations import DeveloperAPI
```
Comment (sven1977) [inline rllib/env/wrappers/unity3d_env.py:10]:
```suggestion
from ray.rllib.utils.annotations import PublicAPI
```

File: rllib/env/wrappers/unity3d_env.py (hunk 7 above)
Comment (sven1977) [inline rllib/env/wrappers/unity3d_env.py:16]:
```suggestion
@PublicAPI
```

File: rllib/utils/exploration/random_encoder.py [hunk 12]
```diff
@@ -16,6 +17,7 @@
...
 class MovingMeanStd:
```
Comment (sven1977) [inline rllib/utils/exploration/random_encoder.py:21]:
```suggestion
class _MovingMeanStd:
```

File: rllib/utils/framework.py (hunk 1 above)
Comment (sven1977) [inline rllib/utils/framework.py:14]:
```suggestion
@PublicAPI
```

File: rllib/utils/framework.py (hunk 2 above)
Comment (sven1977) [inline rllib/utils/framework.py:46]:
```suggestion
@PublicAPI
```

**Review by avnishn — APPROVED** [review]

**Review by gjoliver — APPROVED** [review]

**Pull request merged**